packet_encoder.go

```diff
@@ -50,3 +50,13 @@ type pushEncoder interface {
 	// of data to the saved offset, based on the data between the saved offset and curOffset.
 	run(curOffset int, buf []byte) error
 }
+
+// dynamicPushEncoder extends the interface of pushEncoder for uses cases where the length of the
+// fields itself is unknown until its value was computed (for instance varint encoded lenght
+// fields).
+type dynamicPushEncoder interface {
+	pushEncoder
+
+	// Called during pop() to adjust the length of the field.
+	adjustLength(currOffset int) int
```

Comment:
That will complicate the caller, since in practice most of the time they will call `reserveLength` at `push` time and then again at `pop` time just to readjust.
The initial call to `reserveLength` might not return 0 (especially if we go with your suggestion of computing the size on demand). So, most of the time at the adjust time you want the diff.
I will update the comment.